src/Microsoft.OData.Core/ODataAsynchronousResponseMessage.cs

```diff
@@ -164,9 +206,21 @@ public Stream GetStream()
 
         /// <summary>Asynchronously get the stream backing for this message.</summary>
         /// <returns>The stream backing for this message.</returns>
-        public Task<Stream> GetStreamAsync()
+        public async Task<Stream> GetStreamAsync()
         {
-            return Task<Stream>.Factory.StartNew(this.GetStream);
+            // If writing response, the envelope for the inner message should be written once and only once before returning the stream.
+            if (this.writing && !this.envelopeWritten)
+            {
+                if (this.writeEnvelopeAsync != null)
+                {
+                    await this.writeEnvelopeAsync(this)
+                        .ConfigureAwait(false);
+                }
+
+                this.envelopeWritten = true;

```

Comment:
Out of curiosity, what if `writeEnvelopeAsync` is null but `writeEnvelope` is not null. With this change, the envelope won't be return. But in the current version, it would be written since it calls `GetStream`. With that in mind, isn't it it a breaking change to not fall back to the sync implementation if `writeEnvelope` exists?

Reply:
`ODataAsynchronousResponseMessage` public* API comprises of:
- `StatusCode` property
- `Headers` property (readonly)
- `Container` property (readonly)
- `GetHeader(string)`
- `SetHeader(string, string)`
- `GetStream()`
- `GetStreamAsync()`
- `CreateMessageForWriting(Stream, Action<ODataAsynchronousResponseMessage>, IServiceProvider)` - internal
- `CreateMessageForWritingAsync(Stream, Func<ODataAsynchronousResponseMessage, Task>, IServiceProvider)` - internal
- `CreateMessageForReading(Stream, statusCode, IDictionary<string, string>, IServiceProvider)` - internal

`ODataAsynchronousWriter` public* API on the other hand comprises of:
- `ODataAsynchronousWriter(ODataRawOutputContext)` - internal constructor
- `CreateResponseMessage()`
- `CreateResponseMessageAsync()`
- `Flush()`
- `FlushAsync()`

**Before my change...**
The `ODataAsynchronousResponseMessage.writeEnvelope` delegate is indirectly initialized through either of the following call sequence:

`ODataAsynchronousWriter.CreateResponseMessage` [public] => `ODataAsynchronousWriter.CreateResponseMessageImplementation` [private] => `ODataAsynchronousResponseMessage.CreateMessageForWriting` [interna] => `ODataAsynchronousResponseMessage` [private constructor]

`ODataAsynchronousWriter.CreateResponseMessageAsync` [public] => `ODataAsynchronousWriter.CreateResponseMessageImplementation` [private] => `ODataAsynchronousResponseMessage.CreateMessageForWriting` [interna] => `ODataAsynchronousResponseMessage` [private constructor]

The `writeEnvelope` delegate is passed when `ODataAsynchronousResponseMessage.CreateMessageForWriting` is triggered and this internal method is only triggered from `ODataAsynchronousWriter.CreateResponseMessageImplementation` currently.

**After my change...**
The `ODataAsynchronousResponseMessage.writeEnvelope` delegate will be indirectly initialized through the following call sequence:

`ODataAsynchronousWriter.CreateResponseMessage` [public] => `ODataAsynchronousWriter.CreateResponseMessageImplementation` [private] => `ODataAsynchronousResponseMessage.CreateMessageForWriting` [interna] => `ODataAsynchronousResponseMessage` [private constructor]

`ODataAsynchronousResponseMessage.writeEnvelopeAsync` delegate will be indirectly initialized through the following call sequence:

`ODataAsynchronousWriter.CreateResponseMessageAsync` [public] => `ODataAsynchronousWriter.CreateResponseMessageImplementationAsync` [private] => `ODataAsynchronousResponseMessage.CreateMessageForWritingAsync` [interna] => `ODataAsynchronousResponseMessage` [private constructor]

Since synchronous writing involves using `ODataAsynchronousWriter.CreateResponseMessage`/`ODataAsynchronousResponseMessage.GetStream` pair, while asynchronous writing involves using `ODataAsynchronousWriter.CreateResponseMessageAsync`/`ODataAsynchronousResponseMessage.GetStreamAsync` pair, my change is not a breaking one since `writeEnvelope` will always be null when one is writing asynchronously after I merge my change.

\* _I classified `internal` methods as public in this context so it'd make the explanation easier_

Reply:
Note also that usages of `ODataAsynchronousResponseMessage` are only in `ODataAsynchronousWriter` and `ODataAsynchronousReader` both in ODL. Furthermore, `ODataAsynchronousReader` doesn't make use of the `writeEnvelope` delegate at all.